compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -815,18 +816,23 @@ public BSymbol lookupLangLibMethod(BType type, Name name, SymbolEnv env) {
      * Recursively analyse the symbol env to find the closure variable symbol that is being resolved.
      *
      * @param env       symbol env to analyse and find the closure variable.
+     * @param symbol    symbol to lookup
      * @param name      name of the symbol to lookup
-     * @param expSymTag symbol tag
      * @return closure symbol wrapper along with the resolved count
      */
-    public BSymbol lookupClosureVarSymbol(SymbolEnv env, Name name, long expSymTag) {
+    public BSymbol lookupClosureVarSymbol(SymbolEnv env, Symbol symbol, Name name) {
         ScopeEntry entry = env.scope.lookup(name);
         while (entry != NOT_FOUND_ENTRY) {
+            if (entry.symbol != symbol) {
+                entry = entry.next;
+                continue;
+            }
             if (symTable.rootPkgSymbol.pkgID.equals(entry.symbol.pkgID) &&
                     (entry.symbol.tag & SymTag.VARIABLE_NAME) == SymTag.VARIABLE_NAME) {
                 return entry.symbol;
             }
-            if ((entry.symbol.tag & expSymTag) == expSymTag && !isFieldRefFromWithinARecord(entry.symbol, env)) {
+            if ((entry.symbol.tag & SymTag.VARIABLE) == SymTag.VARIABLE &&
+                    !isFieldRefFromWithinARecord(entry.symbol, env)) {
                 return entry.symbol;
             }
```

Comment:
Since we are checking with the symbol, we can eliminate these steps. Therefore, we can refactor the function as follows

```java
public BSymbol lookupClosureVarSymbol(SymbolEnv env, BSymbol symbol) {
    ScopeEntry entry = env.scope.lookup(symbol.name);
    while (entry != NOT_FOUND_ENTRY) {
        if (entry.symbol == symbol) {
            return entry.symbol;
        }
        entry = entry.next;
    }

    if (env.enclEnv == null) {
        return symTable.notFoundSymbol;
    }

    return lookupClosureVarSymbol(env.enclEnv, symbol);
}
```